src/msg/async/ProtocolV2.cc

```diff
@@ -2464,6 +2464,10 @@ CtPtr ProtocolV2::handle_reconnect(ceph::bufferlist &payload)
     ldout(cct, 1) << __func__
                   << " existing racing replace happened while replacing."
                   << " existing=" << existing << dendl;
+    existing->lock.unlock();
+    // make sure we notice if existing connection is no longer functioning
+    existing->send_keepalive();
+    existing->lock.lock();
```

Comment:
I think both of these call sites dropping lock will introduce bugs.. any time the lock is dropped we need to reverify the connection state.

Instead, can we make an locked variant of send_keepalive?

Reply:
@liewegas I copied these code from https://github.com/ceph/ceph/pull/26668/files#diff-ea445419e9f7f9529b73ce454e5e3c53R2657

But now I am getting confused. Why do we need to drop the lock before calling 
`send_keepalive()`? Can't we simply call `send_keepalive()` under the scope of lock?

Reply:
Ah, it looks like it does work fine with the lock held (it only takes write_lock).  But i suspect the other call site is broken in the same way, then (it shouldn't drop the lock!).  Can you fix it as well?

Reply:
@liewegas Sure. Fixed!